`mapDispatch` docs (#1066)
reduxjs/react-redux: docs/using-react-redux/connect-dispatching-actions-with-mapDispatchToProps.md

---
id: connect-dispatching-actions-with-mapDispatchToProps
title: Connect: Dispatching Actions with mapDispatchToProps
hide_title: true
sidebar_label: Connect: Dispatching Actions with mapStateToProps
---

# Connect: Dispatching Actions with `mapDispatchToProps`

As the second argument passed in to `connect`, `mapDispatchToProps` is used for dispatching actions to the store.

`dispatch` is a function of the Redux store. You call `store.dispatch` to dispatch an action.
This is the only way to trigger a state change.

With React-Redux, your components never access the store directly - `connect` does it for you.
React-Redux gives you two ways to let components dispatch actions:

- By default, a connected component receives `props.dispatch` and can dispatch actions itself.
- `connect` can accept an argument called `mapDispatchToProps`, which lets you create functions that dispatch when called, and pass those functions as props to your component.

> The `mapDispatchToProps` functions are normally referred to as `mapDispatch` for short, and of course the actual variable name doesn’t matter in real code.

Reviewer: ```suggestion
The `mapDispatchToProps` functions are normally referred to as `mapDispatch` for short, but the actual variable name used can be whatever you want.
```

> ## With or Without `mapDispatchToProps`

Reviewer: ```suggestion
## Approaches for Dispatching
```

### Default: `dispatch` as a Prop

If you don't specify the second argument to `connect()`, your component will receive `dispatch` by default. For example:

```js
connect()(MyComponent);
// which is equivalent with
connect(
  null,
  null
)(MyComponent);

// or
connect(mapStateToProps /** no second argument */)(MyComponent);
```

Once you have connected your component in this way, your component receives `props.dispatch`. You may use it to dispatch actions to the store.

```js
function Counter({ count, dispatch }) {
  return (
    <div>
      <button onClick={() => dispatch({ type: "DECREMENT" })}>-</button>
      <span>{count}</span>
      <button onClick={() => dispatch({ type: "INCREMENT" })}>+</button>
      <button onClick={() => dispatch({ type: "RESET" })}>reset</button>
    </div>
  );
}
```

> ### Providing A `mapDispatchToProps`

Reviewer: ```suggestion
### Providing A `mapDispatchToProps` Parameter
```

Providing a `mapDispatchToProps` allows you to specify which actions your component might need to dispatch. It lets you provide action dispatching functions as props. Therefore, instead of calling `props.dispatch(() => increment())`, you may call `props.increment()` directly. There are a few reasons why you might want to do that.

#### More Declarative

First, encapsulating the dispatch logic into function makes the implementation more declarative.

> Dispatching an action and let the Redux store handle the data flow is _how to_ implement the feature, it is not _what_ it does.

Reviewer: ```suggestion
Dispatching an action and letting the Redux store handle the data flow is _how to_ implement the behavior, rather than _what_ it does.
```

Therefore, instead of saying:

```js
<button onClick={() => dispatch({ type: "SOMETHING" })} />
```

you may prefer saying:

```js
<button onClick={doSomething} />
```

Once you've wrapped all our action creators with functions that dispatch the actions, the component is free of the need of `dispatch`.
Therefore, **if you define your own `mapDispatchToProps`, the connected component will no longer receive `dispatch`.**

#### Pass Down Action Dispatching Logic to ( Unconnected ) Child Components

> Furthermore, you also gain the freedom of passing down the action dispatching functions to child ( likely unconnected ) components. Now, both the connected component and its descendant need not be aware of `dispatch`, while performing their communications with the Redux store.

Reviewer: ```suggestion
In addition, you also gain the ability to pass down the action dispatching functions to child ( likely unconnected ) components. This allows more components to dispatch actions, while keeping them "unaware" of Redux.  A good example would be dispatching an action when a button is clicked.  Connecting the button directly probably doesn't make sense conceptually, and neither does having the button reference `dispatch`.  Instead, we just want to use `<button onClick={increment} />`.
```

Reviewer: Hah, I wrote this suggestion and then went back and saw you'd said basically the same thing in the previous heading :)

Tell you what.  Let's keep the "Declarative" section as just the couple paragraphs, and move all the `<button>` examples into this heading.

Author: What about like this :)

This is what React-Redux’s `connect` does — it encapsulates the logic of talking to the Redux store and lets you not worry about it. And this is what you should totally make full use of in your implementation.

## Two Forms of `mapDispatchToProps`

The `mapDispatchToProps` parameter can be of two forms. While the function form allows more customization, the object form is easy to use.

- **Function form**: Allows more customization, gains access to `dispatch` and optionally `ownProps`

> - **Object short hand form**: More declarative and easier to use

Reviewer: ```suggestion
- **Object shorthand form**: More declarative and easier to use
```

> ⭐ We recommend using the object form of `mapDispatchToProps` unless you specifically need to customize dispatching behavior in some way.

Reviewer: ```suggestion
> ⭐ **Note**: We recommend using the object form of `mapDispatchToProps` unless you specifically need to customize dispatching behavior in some way.
```

## Defining `mapDispatchToProps` As A Function

Defining `mapDispatchToProps` as a function gives you the most flexibility in customizing the functions your component receives, and how they dispatch actions.
You gain access to `dispatch` and `ownProps`.
You may use this chance to write customized functions to be called by your connected components.

### Arguments

1. **`dispatch`**
2. **`ownProps` (optional)**

**`dispatch`**

> The `mapDispatchToProps` function will be called with `dispatch` as the argument. You may use this param to wrap around the actions you wish to dispatch, or the action creators that return the actions.

Reviewer: ```suggestion
The `mapDispatchToProps` function will be called with `dispatch` as the first argument. You will normally make use of this by returning new functions that call `dispatch()` inside themselves, and either pass in a plain action object directly or pass in the result of an action creator.
```

```js
const mapDispatchToProps = dispatch => {
  return {
    // dispatching plain actions
    increment: () => dispatch({ type: "INCREMENT" }),
    decrement: () => dispatch({ type: "DECREMENT" }),
    reset: () => dispatch({ type: "RESET" })
  };
};
```

> You may take this chance to forward arguments to your action creators:

Reviewer: ```suggestion
You will also likely want to forward arguments to your action creators:
```

```js
const mapDispatchToProps = dispatch => {
  return {
    // explicitly forwarding arguments

>     onClick: event => dispatch(trackClick(event)),

Reviewer: GREAT examples here! :)

    // implicitly forwarding arguments
    onReceiveImpressions: (...impressions) =>
      dispatch(trackImpressions(impressions))
  };
};
```

**`ownProps` ( optional )**

If your `mapDispatchToProps` function is declared as taking two parameters, it will be called with `dispatch` as the first parameter and the `props` passed to the connected component as the second parameter, and will be re-invoked whenever the connected component receives new props.

This means, instead of re-binding new `props` to action dispatchers upon component re-rendering, you may do so when your component's `props` change.

```js
// binds on component re-rendering
<button onClick={() => this.props.toggleTodo(this.props.todoId)} />;

// binds on `props` change
const mapDispatchToProps = (dispatch, ownProps) => {
  toggleTodo: () => dispatch(toggleTodo(ownProps.todoId));
};
```

### Return

Your `mapDispatchToProps` function should return a plain object:

> - Each field in the object should be a function, calling which should automatically dispatch the action

Reviewer: ```suggestion
- Each field in the object will become a separate prop for your own component, and the value should normally be a function that dispatches an action when called.
```

```js
function Counter({ count, increment, decrement, reset }) {
  return (
    <div>
      <button onClick={decrement}>-</button>
      <span>{count}</span>
      <button onClick={increment}>+</button>
      <button onClick={reset}>reset</button>
    </div>
  );
}
```

(Full code of the Counter example is [in this CodeSandbox](https://codesandbox.io/s/yv6kqo1yw9))

### Defining the `mapDispatchToProps` Function with `bindActionCreators`

Wrapping these functions by hand is tedious, so Redux provides a function to simplify that.

> `bindActionCreators` turns an object whose values are [action creators](https://redux.js.org/glossary#action-creator), into an object with the same keys, but with every action creator wrapped into a [`dispatch`](https://redux.js.org/api/store#dispatch) call so they may be invoked directly. See [Redux Docs on `bindActionCreators`](http://redux.js.org/docs/api/bindActionCreators.html)

`bindActionCreators` accepts two parameters:

1. A **`function`** (an action creator) or an **`object`** (each field an action creator)
2. `dispatch`

> 

Reviewer: ```suggestion

The wrapper functions generated by `bindActionCreators` will automatically forward all of their arguments, so you don't need to do that by hand.
```

```js
import { bindActionCreators } from "redux";

const increment = () => ({ type: "INCREMENT" });
const decrement = () => ({ type: "DECREMENT" });
const reset = () => ({ type: "RESET" });

// binding an action creator

> // returns () => dispatch(increment())

Reviewer: ```suggestion
// returns (...args) => dispatch(increment(...args))
```

Can maybe skip doing this for the rest of them?

> bindActionCreators(increment, dispatch);

Reviewer: ```suggestion
const boundIncrement = bindActionCreators(increment, dispatch);
```

// binding an object full of action creators

> bindActionCreators(

Reviewer: ```suggestion
const boundActionCreators = bindActionCreators(
```

>   {

Reviewer: Might as well collapse this object to one line

    increment,
    decrement,
    reset
  },
  dispatch
);
// returns
// {

> //   increment: () => dispatch(increment()),

Reviewer: Eh, let's go ahead and show all the `...args` in these comments after all.

//   decrement: () => dispatch(decrement()),
//   reset: () => dispatch(reset()),
// }
```

To use `bindActionCreators` in our `mapDispatchToProps` function:

```js
import { bindActionCreators } from "redux";
// ...

function mapDispatchToProps(dispatch) {
  return bindActionCreators(

>     {

Reviewer: I think we can collapse this object too, and leave off the comments - that point was made already.

      increment, // () => dispatch(increment())
      decrement, // () => dispatch(decrement())
      reset // () => dispatch(reset())
    },
    dispatch
  );
}

// component receives props.increment, props.decrement, props.reset
connect(
  null,
  mapDispatchToProps
)(Counter);
```

### Manually Injecting `dispatch`

> Since the `mapDispatchToProps` argument is supplied, the component will no longer receive the default `dispatch`. You may bring it back by adding it manually to the return of your `mapDispatchToProps`, although most of the time you shouldn’t need to do this:

Reviewer: ```suggestion
If the `mapDispatchToProps` argument is supplied, the component will no longer receive the default `dispatch`. You may bring it back by adding it manually to the return of your `mapDispatchToProps`, although most of the time you shouldn’t need to do this:
```

```js
import { bindActionCreators } from "redux";
// ...

function mapDispatchToProps(dispatch) {
  return {
    dispatch,
    ...bindActionCreators(

>       {

Reviewer: Collapse object.

**We recommend always using the “object shorthand” form of `mapDispatchToProps`, unless you have a specific reason to customize the dispatching behavior.**

Note that:

- Each field of the `mapDispatchToProps` object is assumed to be an action creator
- Your component will no longer receive `dispatch` as a prop

```js
// React-Redux does this for you automatically:
dispatch => bindActionCreators(mapDispatchToProps, dispatch);
```

Therefore, our `mapDispatchToProps` can simply be:

```js
const mapDispatchToProps = {
  increment,
  decrement,
  reset
};
```

Since the actual name of the variable is up to you, you might want to give it a name like `actionCreators`, or even define the object inline in the call to `connect`:

```js
import {increment, decrement, reset} from "./counterActions";

const actionCreators = {
  increment,
  decrement,
  reset
}

export default connect(mapState, actionCreators)(Counter);

// or
export default connect(

>   null,

Reviewer: ```suggestion
  mapState,
```

Might as well be consistent.

Also known as

```js
TypeError: this.props.dispatch is not a function
```

This is a common error that happens when you try to call `this.props.dispatch` , but `dispatch` is not injected to your component.

`dispatch` is injected to your component _only_ when:

**1. You do not provide `mapDispatchToProps`**

The default `mapDispatchToProps` is simply `dispatch => ({ dispatch })`. If you do not provide `mapDispatchToProps`, `dispatch` will be provided as mentioned above.

In another words, if you do:

```js
// component receives `dispatch`
connect(mapStateToProps /** no second argument*/)(Component);
```

**2. Your customized `mapDispatchToProps` function return specifically contains `dispatch`**

You may bring back `dispatch` by providing your customized `mapDispatchToProps` function:

```js
const mapDispatchToProps = dispatch => {
  return {
    increment: () => dispatch(increment()),
    decrement: () => dispatch(decrement()),
    reset: () => dispatch(reset()),
    dispatch
  };
};
```

Or alternatively, with `bindActionCreators`:

```js
import { bindActionCreators } from "redux";

function mapDispatchToProps(dispatch) {
  return {
    dispatch,
    ...bindActionCreators(

>       {

Reviewer: Collapse object :)

(Hey, you were consistent :)  )

        increment,
        decrement,
        reset
      },
      dispatch
    )
  };
}
```

See [this error in action in Redux’s GitHub issue #255](https://github.com/reduxjs/react-redux/issues/255).

There are discussions regarding whether to provide `dispatch` to your components when you specify `mapDispatchToProps` ( [Dan Abramov’s response to #255](https://github.com/reduxjs/react-redux/issues/255#issuecomment-172089874) ). You may read them for further understanding of the current implementation intention.

### Can I `mapDispatchToProps` without `mapStateToProps` in Redux?

> Yes. You can skip the first parameter by passing `undefined` or `null`. Your component will not subscribe to the store, and gain dispatch props defined by `mapStateToProps`.

Reviewer: ```suggestion
Yes. You can skip the first parameter by passing `undefined` or `null`. Your component will not subscribe to the store, and will still receive the dispatch props defined by `mapStateToProps`.
```

```js
connect(
  null,
  mapDispatchToProps
)(MyComponent);
```

### Can I call `store.dispatch`?

> It is an anti-pattern to use `store.dispatch` which you may retrieve from the `store` in context. As explained in [Redux's FAQ on store setup](https://redux.js.org/faq/storesetup#can-or-should-i-create-multiple-stores-can-i-import-my-store-directly-and-use-it-in-components-myself), any direct access of the store in a component is not recommended, whether it be via direct import or from context. Let React-Redux’s `connect` handle the access to the store, and use the `dispatch` it passes to the props to dispatch actions.

Reviewer: ```suggestion
It's an anti-pattern to interact with the store directly in a React component, whether it's an explicit import of the store or accessing it via context (see the [Redux FAQ entry on store setup](https://redux.js.org/faq/storesetup#can-or-should-i-create-multiple-stores-can-i-import-my-store-directly-and-use-it-in-components-myself) for more details). Let React-Redux’s `connect` handle the access to the store, and use the `dispatch` it passes to the props to dispatch actions.
```
